File: tools/extended_devices/dishwasher.py
from __future__ import annotations

from ..base import BaseTool, ToolResult
from ..schemas.extended_devices import DishwasherInput


class DishwasherTool(BaseTool):
    name = "dishwasher_control"
    description = "식기세척기를 제어합니다. 가동, 일시정지, 정지 및 세척 모드를 설정할 수 있습니다."
    args_schema = DishwasherInput

    def execute(self, params: DishwasherInput) -> ToolResult:
        data = self.client.command(
            params.device_id,
            "dishwasherOperatingState",
            "setMachineState",
            arguments=[params.action],
        )
        if "error" in data:
            error = data.get("error", {})
            message = (
                error.get("message", str(data))
                if isinstance(error, dict)
                else str(error)
            )
            return ToolResult(success=False, error=message)

        if params.mode:
            data = self.client.command(
                params.device_id,
                "dishwasherMode",
                "setDishwasherMode",
                arguments=[params.mode],
            )
            if "error" in data:
                error = data.get("error", {})
                message = (
                    error.get("message", str(data))
                    if isinstance(error, dict)
                    else str(error)
                )
                return ToolResult(success=False, error=message)

        return ToolResult(success=True, data=data)

File: smartthings_tools_rs/src/extended_devices/dishwasher.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::extended_devices::DishwasherInput;
use crate::tool::{Tool, ToolResult};

pub struct DishwasherTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for DishwasherTool<'_> {
    fn name(&self) -> &'static str {
        "dishwasher_control"
    }
    fn description(&self) -> &'static str {
        "식기세척기를 제어합니다. 가동, 일시정지, 정지 및 세척 모드를 설정할 수 있습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(DishwasherInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: DishwasherInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };
        let data = self.client.command(
            &input.device_id,
            "dishwasherOperatingState",
            "setMachineState",
            Some(vec![Value::String(input.action.clone())]),
            None,
        );
        if has_error(&data) {
            return ToolResult::fail(extract_error_msg(&data));
        }

        if let Some(ref mode) = input.mode {
            let data = self.client.command(
                &input.device_id,
                "dishwasherMode",
                "setDishwasherMode",
                Some(vec![Value::String(mode.clone())]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
        }

        ToolResult::ok(serde_json::to_value(&data).unwrap_or_default())
    }
}
